ln-simln-jamming/src/parsing.rs

```diff
@@ -181,96 +184,199 @@ fn parse_duration(s: &str) -> Result<Duration, String> {
+        let mut end = current_breakpoint + chunk_size;
+        if end > file_size {
+            end = file_size;
+        }
+
+        breakpoints.push(current_breakpoint);
+        let next_eol_point = find_next_newline(&mut reader, end)?;
+        current_breakpoint = next_eol_point + 1;
+    }
+
+    Ok(breakpoints)
+}
+
 /// Reads forwards from a CSV (generated by simln), optionally filtering to only get a set duration of forwards from
 /// the file.
-pub fn history_from_file(
+pub async fn history_from_file(
     file_path: &PathBuf,
     filter_duration: Option<Duration>,
 ) -> Result<Vec<BootstrapForward>, BoxError> {
-    let file = File::open(file_path)?;
-    let reader = BufReader::new(file);
-    let mut csv_reader = csv::Reader::from_reader(reader);
+    let num_chunks = Handle::current().metrics().num_workers().div_ceil(2);
+    let breakpoints = calc_file_chunks(file_path, num_chunks as u8)?;
 
-    let mut forwards = Vec::new();
-    let mut start_ts = None;
-
-    for result in csv_reader.records() {
-        let record: StringRecord = result?;
-
-        // We can skip 6/7 because they're outgoing timestamps, we only care about when the htlc is fully removed
-        // from the incoming link (for simplicity's sake).
-        let incoming_amt: u64 = record[0].parse()?;
-        let incoming_expiry: u32 = record[1].parse()?;
-        let incoming_add_ts: u64 = record[2].parse()?;
-        let incoming_remove_ts: u64 = record[3].parse()?;
-        let outgoing_amt: u64 = record[4].parse()?;
-        let outgoing_expiry: u32 = record[5].parse()?;
-        let forwarding_node = PublicKey::from_slice(&hex::decode(&record[8])?)?;
-        let channel_in_id: u64 = record[10].parse()?;
-        let channel_out_id: u64 = record[11].parse()?;
-
-        // If we're filtering cut off any htlc that was in flight at the cutoff point.
+    let file = File::open(file_path)?;
+    let file_size = file.metadata()?.len();
+    let filter_cutoff = {
         if let Some(duration) = filter_duration {
-            let cutoff = match start_ts {
-                Some(s) => s,
-                None => {
-                    start_ts = Some(incoming_add_ts);
-                    incoming_add_ts
+            let reader = BufReader::new(file);
+            let mut csv_reader = csv::Reader::from_reader(reader);
+            let mut first_record = StringRecord::new();
+            csv_reader.read_record(&mut first_record)?;
+            let incoming_add_ts: u64 = first_record[2].parse()?;
+            Some(incoming_add_ts.add(duration.as_nanos() as u64))
+        } else {
+            None
+        }
+    };
+
+    let mut tasks: Vec<tokio::task::JoinHandle<Result<Vec<BootstrapForward>, BoxError>>> =
+        Vec::with_capacity(num_chunks);
+
+    for i in 0..num_chunks {
+        let start = breakpoints[i];
+        let end = if i == num_chunks - 1 {
+            file_size
+        } else {
+            breakpoints[i + 1]
+        };
```

Comment:
Related to the above comment about how we create breakpoints, could this be simplified to just be `0...breakpoints.len() -1`? So that if we need fewer chunks than we thought, we just dispatch the number that we want?